crates/shim-sgx/src/handler/mod.rs

```diff
@@ -230,9 +230,46 @@ impl guest::Handler for Handler<'_> {
     fn munmap(
         &mut self,
         _platform: &impl Platform,
-        _addr: NonNull<c_void>,
-        _length: c_size_t,
+        addr: NonNull<c_void>,
+        length: c_size_t,
     ) -> sallyport::Result<()> {
+        let addr = addr.as_ptr() as usize;
+        let length = (length + 0xfff) & !0xfff;
+
+        if addr & 0xfff != 0 || length == 0 {
+            return Err(EINVAL);
+        }
+
+        let mut heap = HEAP.write();
+
+        let region = Region::new(Address::new(addr), Address::new(addr + length));
+
+        if heap.contains(region) == None {
+            return Ok(());
+        }
+
+        // Process the ledger first, before doing anything else, because it can
+        // legitly fail when running out of resources.
+        if heap.munmap(region).is_err() {
+            return Err(ENOMEM);
+        }
+
+        // On the other hand, failing in any of these operations is expected to
+        // crash the enclave because it is due either to a software bug, or a
+        // malicious host.
+        self.trim_sgx_pages(NonNull::new(addr as *mut _).unwrap(), length)?;
```

Comment:
In the case of a malicious host, you should call `self.attacked()` rather than crashing the enclave, if possible.

Reply:
fixed